### Prediction between Downtime and Availability(Makino - 4) using Random Forest and Decision tree Algorithms

#### The MSE score provides valuable insights into the predictive accuracy and performance of your regression model

## Random forest

In [2]:
# Import libraries
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

#scaler = MinMaxScaler()  # create an instance of the MinMaxScaler class

# Fit the scaler on the data and transform the data
#X_scaled = scaler.fit_transform(X)


In [21]:
Makino_data = pd.read_csv(r"C:\Users\15878\DA\SGD_Data.csv")

Makino_data

Date En_Name   Bg_Name Plant_Name Line_Name Machine_Name   Shift  \
0     01-01-2023    RANE  ZRAI-SGD    Plant 2     7 LPM     Makino 4  Shift1   
1     01-01-2023    RANE  ZRAI-SGD    Plant 2     7 LPM     Makino 4  Shift2   
2     01-01-2023    RANE  ZRAI-SGD    Plant 2     7 LPM     Makino 4  Shift2   
3     01-01-2023    RANE  ZRAI-SGD    Plant 2     7 LPM     Makino 4  Shift3   
4     01-01-2023    RANE  ZRAI-SGD    Plant 2     7 LPM     Makino 4  Shift3   
...          ...     ...       ...        ...       ...          ...     ...   
1205  30-12-2022    RANE  ZRAI-SGD    Plant 2     7 LPM     Makino 4  Shift3   
1206  30-12-2022    RANE  ZRAI-SGD    Plant 2     7 LPM     Makino 4  Shift3   
1207  31-12-2022    RANE  ZRAI-SGD    Plant 2     7 LPM     Makino 4  Shift1   
1208  31-12-2022    RANE  ZRAI-SGD    Plant 2     7 LPM     Makino 4  Shift2   
1209  31-12-2022    RANE  ZRAI-SGD    Plant 2     7 LPM     Makino 4  Shift3   

      idle_Cycle_time         HT          MT  ...  Run_Time  Down_Time  \
0               145.0   0.000000    0.000000  ...         0        509   
1                 0.0   0.000000    0.000000  ...         0        270   
2               145.0   0.000000    0.000000  ...         0        208   
3                 0.0   0.000000    0.000000  ...         0         30   
4               145.0  27.500000  119.277778  ...       196        144   
...               ...        ...         ...  ...       ...        ...   
1205            145.0  42.000000   90.000000  ...        14         16   
1206            150.0  49.269231  130.730769  ...       314          8   
1207            145.0  39.088235  176.058823  ...       324        109   
1208            145.0  47.500000  117.576923  ...       241        199   
1209            145.0  63.071429  143.392857  ...       263         62   

      Total_Count  rejection  Good_Count  Operating_Time     Quality  \
0               0          0           0             509    0.000000   
1               0          0           0             270    0.000000   
2               0          0           0             208    0.000000   
3               0          0           0              30    0.000000   
4              72          4          68             358   94.444444   
...           ...        ...         ...             ...         ...   
1205            4          0           4              31  100.000000   
1206          118          4         114             340   96.610169   
1207          122          8         114             466   93.442623   
1208           98          8          90             465   91.836735   
1209          102          0         102             378  100.000000   

      Availability  Performance        OEE  
0         0.000000     0.000000   0.000000  
1         0.000000     0.000000   0.000000  
2         0.000000     0.000000   0.000000  
3         0.000000     0.000000   0.000000  
4        60.007446    80.961613  45.909578  
...            ...          ...        ...  
1205     47.120709    68.155112  32.115172  
1206     97.775486    88.895585  84.021226  
1207     70.332229    82.490091  59.137092  
1208     57.212986    88.996054  46.809426  
1209     83.666166    78.101072  65.344172  

[1210 rows x 21 columns]

In [4]:
# Load data
data = pd.read_csv(r"C:\Users\15878\DA\SGD_Data_Pred.csv")

data




idle_Cycle_time         HT          MT  Stop_time  Run_Time  Down_Time  \
0               145.0   0.000000    0.000000          0         0        509   
1                 0.0   0.000000    0.000000          0         0        270   
2               145.0   0.000000    0.000000          0         0        208   
3                 0.0   0.000000    0.000000          0         0         30   
4               145.0  27.500000  119.277778         20       196        144   
...               ...        ...         ...        ...       ...        ...   
1205            145.0  42.000000   90.000000          1        14         16   
1206            150.0  49.269231  130.730769         19       314          8   
1207            145.0  39.088235  176.058823         35       324        109   
1208            145.0  47.500000  117.576923         26       241        199   
1209            145.0  63.071429  143.392857         54       263         62   

      Total_Count  rejection  Good_Count  Operating_Time     Quality  \
0               0          0           0             509    0.000000   
1               0          0           0             270    0.000000   
2               0          0           0             208    0.000000   
3               0          0           0              30    0.000000   
4              72          4          68             358   94.444444   
...           ...        ...         ...             ...         ...   
1205            4          0           4              31  100.000000   
1206          118          4         114             340   96.610169   
1207          122          8         114             466   93.442623   
1208           98          8          90             465   91.836735   
1209          102          0         102             378  100.000000   

      Availability  Performance        OEE  
0         0.000000     0.000000   0.000000  
1         0.000000     0.000000   0.000000  
2         0.000000     0.000000   0.000000  
3         0.000000     0.000000   0.000000  
4        60.007446    80.961613  45.909578  
...            ...          ...        ...  
1205     47.120709    68.155112  32.115172  
1206     97.775486    88.895585  84.021226  
1207     70.332229    82.490091  59.137092  
1208     57.212986    88.996054  46.809426  
1209     83.666166    78.101072  65.344172  

[1210 rows x 14 columns]

In [5]:
data.columns

Index(['idle_Cycle_time', 'HT', 'MT', 'Stop_time', 'Run_Time', 'Down_Time',
       'Total_Count', 'rejection', 'Good_Count', 'Operating_Time', 'Quality',
       'Availability', 'Performance', 'OEE'],
      dtype='object')

In [6]:
data_arr=data.values
data_arr

array([[145.        ,   0.        ,   0.        , ...,   0.        ,
          0.        ,   0.        ],
       [  0.        ,   0.        ,   0.        , ...,   0.        ,
          0.        ,   0.        ],
       [145.        ,   0.        ,   0.        , ...,   0.        ,
          0.        ,   0.        ],
       ...,
       [145.        ,  39.08823529, 176.0588235 , ...,  70.33222918,
         82.49009093,  59.13709216],
       [145.        ,  47.5       , 117.5769231 , ...,  57.21298552,
         88.99605436,  46.80942645],
       [145.        ,  63.07142857, 143.3928571 , ...,  83.66616595,
         78.10107198,  65.34417248]])

In [7]:
b = np.isinf(data_arr)
b

array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False]])

## Data Cleaning

In [8]:
data.dropna()
data

idle_Cycle_time         HT          MT  Stop_time  Run_Time  Down_Time  \
0               145.0   0.000000    0.000000          0         0        509   
1                 0.0   0.000000    0.000000          0         0        270   
2               145.0   0.000000    0.000000          0         0        208   
3                 0.0   0.000000    0.000000          0         0         30   
4               145.0  27.500000  119.277778         20       196        144   
...               ...        ...         ...        ...       ...        ...   
1205            145.0  42.000000   90.000000          1        14         16   
1206            150.0  49.269231  130.730769         19       314          8   
1207            145.0  39.088235  176.058823         35       324        109   
1208            145.0  47.500000  117.576923         26       241        199   
1209            145.0  63.071429  143.392857         54       263         62   

      Total_Count  rejection  Good_Count  Operating_Time     Quality  \
0               0          0           0             509    0.000000   
1               0          0           0             270    0.000000   
2               0          0           0             208    0.000000   
3               0          0           0              30    0.000000   
4              72          4          68             358   94.444444   
...           ...        ...         ...             ...         ...   
1205            4          0           4              31  100.000000   
1206          118          4         114             340   96.610169   
1207          122          8         114             466   93.442623   
1208           98          8          90             465   91.836735   
1209          102          0         102             378  100.000000   

      Availability  Performance        OEE  
0         0.000000     0.000000   0.000000  
1         0.000000     0.000000   0.000000  
2         0.000000     0.000000   0.000000  
3         0.000000     0.000000   0.000000  
4        60.007446    80.961613  45.909578  
...            ...          ...        ...  
1205     47.120709    68.155112  32.115172  
1206     97.775486    88.895585  84.021226  
1207     70.332229    82.490091  59.137092  
1208     57.212986    88.996054  46.809426  
1209     83.666166    78.101072  65.344172  

[1210 rows x 14 columns]

In [9]:
data[['HT', 'MT', 'Quality','Availability','Performance','OEE']] = data[['HT', 'MT', 'Quality','Availability','Performance','OEE']].round(2)
data

idle_Cycle_time     HT      MT  Stop_time  Run_Time  Down_Time  \
0               145.0   0.00    0.00          0         0        509   
1                 0.0   0.00    0.00          0         0        270   
2               145.0   0.00    0.00          0         0        208   
3                 0.0   0.00    0.00          0         0         30   
4               145.0  27.50  119.28         20       196        144   
...               ...    ...     ...        ...       ...        ...   
1205            145.0  42.00   90.00          1        14         16   
1206            150.0  49.27  130.73         19       314          8   
1207            145.0  39.09  176.06         35       324        109   
1208            145.0  47.50  117.58         26       241        199   
1209            145.0  63.07  143.39         54       263         62   

      Total_Count  rejection  Good_Count  Operating_Time  Quality  \
0               0          0           0             509     0.00   
1               0          0           0             270     0.00   
2               0          0           0             208     0.00   
3               0          0           0              30     0.00   
4              72          4          68             358    94.44   
...           ...        ...         ...             ...      ...   
1205            4          0           4              31   100.00   
1206          118          4         114             340    96.61   
1207          122          8         114             466    93.44   
1208           98          8          90             465    91.84   
1209          102          0         102             378   100.00   

      Availability  Performance    OEE  
0             0.00         0.00   0.00  
1             0.00         0.00   0.00  
2             0.00         0.00   0.00  
3             0.00         0.00   0.00  
4            60.01        80.96  45.91  
...            ...          ...    ...  
1205         47.12        68.16  32.12  
1206         97.78        88.90  84.02  
1207         70.33        82.49  59.14  
1208         57.21        89.00  46.81  
1209         83.67        78.10  65.34  

[1210 rows x 14 columns]

In [10]:
data.fillna(0)
data

idle_Cycle_time     HT      MT  Stop_time  Run_Time  Down_Time  \
0               145.0   0.00    0.00          0         0        509   
1                 0.0   0.00    0.00          0         0        270   
2               145.0   0.00    0.00          0         0        208   
3                 0.0   0.00    0.00          0         0         30   
4               145.0  27.50  119.28         20       196        144   
...               ...    ...     ...        ...       ...        ...   
1205            145.0  42.00   90.00          1        14         16   
1206            150.0  49.27  130.73         19       314          8   
1207            145.0  39.09  176.06         35       324        109   
1208            145.0  47.50  117.58         26       241        199   
1209            145.0  63.07  143.39         54       263         62   

      Total_Count  rejection  Good_Count  Operating_Time  Quality  \
0               0          0           0             509     0.00   
1               0          0           0             270     0.00   
2               0          0           0             208     0.00   
3               0          0           0              30     0.00   
4              72          4          68             358    94.44   
...           ...        ...         ...             ...      ...   
1205            4          0           4              31   100.00   
1206          118          4         114             340    96.61   
1207          122          8         114             466    93.44   
1208           98          8          90             465    91.84   
1209          102          0         102             378   100.00   

      Availability  Performance    OEE  
0             0.00         0.00   0.00  
1             0.00         0.00   0.00  
2             0.00         0.00   0.00  
3             0.00         0.00   0.00  
4            60.01        80.96  45.91  
...            ...          ...    ...  
1205         47.12        68.16  32.12  
1206         97.78        88.90  84.02  
1207         70.33        82.49  59.14  
1208         57.21        89.00  46.81  
1209         83.67        78.10  65.34  

[1210 rows x 14 columns]

In [11]:
data.dtypes

idle_Cycle_time    float64
HT                 float64
MT                 float64
Stop_time            int64
Run_Time             int64
Down_Time            int64
Total_Count          int64
rejection            int64
Good_Count           int64
Operating_Time       int64
Quality            float64
Availability       float64
Performance        float64
OEE                float64
dtype: object

In [12]:
data.count()

idle_Cycle_time    1201
HT                 1210
MT                 1210
Stop_time          1210
Run_Time           1210
Down_Time          1210
Total_Count        1210
rejection          1210
Good_Count         1210
Operating_Time     1210
Quality            1210
Availability       1210
Performance        1210
OEE                1210
dtype: int64

### Data Preprocessing

####  Preprocess the data to remove outliers, handle missing values, and scale the features appropriately. This can help the model better learn the underlying patterns in the data.

In [13]:
# Load data
data = pd.read_csv(r"C:\Users\15878\DA\SGD_Data_Pred.csv")

data

idle_Cycle_time         HT          MT  Stop_time  Run_Time  Down_Time  \
0               145.0   0.000000    0.000000          0         0        509   
1                 0.0   0.000000    0.000000          0         0        270   
2               145.0   0.000000    0.000000          0         0        208   
3                 0.0   0.000000    0.000000          0         0         30   
4               145.0  27.500000  119.277778         20       196        144   
...               ...        ...         ...        ...       ...        ...   
1205            145.0  42.000000   90.000000          1        14         16   
1206            150.0  49.269231  130.730769         19       314          8   
1207            145.0  39.088235  176.058823         35       324        109   
1208            145.0  47.500000  117.576923         26       241        199   
1209            145.0  63.071429  143.392857         54       263         62   

      Total_Count  rejection  Good_Count  Operating_Time     Quality  \
0               0          0           0             509    0.000000   
1               0          0           0             270    0.000000   
2               0          0           0             208    0.000000   
3               0          0           0              30    0.000000   
4              72          4          68             358   94.444444   
...           ...        ...         ...             ...         ...   
1205            4          0           4              31  100.000000   
1206          118          4         114             340   96.610169   
1207          122          8         114             466   93.442623   
1208           98          8          90             465   91.836735   
1209          102          0         102             378  100.000000   

      Availability  Performance        OEE  
0         0.000000     0.000000   0.000000  
1         0.000000     0.000000   0.000000  
2         0.000000     0.000000   0.000000  
3         0.000000     0.000000   0.000000  
4        60.007446    80.961613  45.909578  
...            ...          ...        ...  
1205     47.120709    68.155112  32.115172  
1206     97.775486    88.895585  84.021226  
1207     70.332229    82.490091  59.137092  
1208     57.212986    88.996054  46.809426  
1209     83.666166    78.101072  65.344172  

[1210 rows x 14 columns]

In [6]:
from sklearn.preprocessing import MinMaxScaler
data = pd.read_csv(r"C:\Users\15878\DA\SGD_Data_Pred.csv")
# create a MinMaxScaler object
scaler = MinMaxScaler()

# select columns to scale
cols_to_scale = ['Availability']
cols_to_scale1 = ['Down_Time']

# apply Min-Max scaling to selected columns
data[cols_to_scale] = scaler.fit_transform(data[cols_to_scale])
data[cols_to_scale1] = scaler.fit_transform(data[cols_to_scale1])
print(data[cols_to_scale],data[cols_to_scale1])



      Availability
0         0.000000
1         0.000000
2         0.000000
3         0.000000
4         0.600074
...            ...
1205      0.471207
1206      0.977755
1207      0.703322
1208      0.572130
1209      0.836662

[1210 rows x 1 columns]       Down_Time
0      0.990272
1      0.525292
2      0.404669
3      0.058366
4      0.280156
...         ...
1205   0.031128
1206   0.015564
1207   0.212062
1208   0.387160
1209   0.120623

[1210 rows x 1 columns]


In [14]:
from sklearn.preprocessing import MinMaxScaler
data = pd.read_csv(r"C:\Users\15878\DA\SGD_Data_Pred.csv")
# create a MinMaxScaler object
scaler = MinMaxScaler()

# select columns to scale
cols_to_scale = ['Down_Time']
cols_to_scale1 = ['Availability']

# apply Min-Max scaling to selected columns
data[cols_to_scale] = scaler.fit_transform(data[cols_to_scale])
data[cols_to_scale1] = scaler.fit_transform(data[cols_to_scale1])
print(data[cols_to_scale],data[cols_to_scale1])

      Down_Time
0      0.990272
1      0.525292
2      0.404669
3      0.058366
4      0.280156
...         ...
1205   0.031128
1206   0.015564
1207   0.212062
1208   0.387160
1209   0.120623

[1210 rows x 1 columns]       Availability
0         0.000000
1         0.000000
2         0.000000
3         0.000000
4         0.600074
...            ...
1205      0.471207
1206      0.977755
1207      0.703322
1208      0.572130
1209      0.836662

[1210 rows x 1 columns]


In [15]:
# Split data into features and target variable
X = data[cols_to_scale]
y = data[cols_to_scale1]
print(X,y)

      Down_Time
0      0.990272
1      0.525292
2      0.404669
3      0.058366
4      0.280156
...         ...
1205   0.031128
1206   0.015564
1207   0.212062
1208   0.387160
1209   0.120623

[1210 rows x 1 columns]       Availability
0         0.000000
1         0.000000
2         0.000000
3         0.000000
4         0.600074
...            ...
1205      0.471207
1206      0.977755
1207      0.703322
1208      0.572130
1209      0.836662

[1210 rows x 1 columns]


In [7]:
# Split data into features and target variable
X = data[cols_to_scale]
y = data[cols_to_scale1]
print(X,y)

      Availability
0         0.000000
1         0.000000
2         0.000000
3         0.000000
4         0.600074
...            ...
1205      0.471207
1206      0.977755
1207      0.703322
1208      0.572130
1209      0.836662

[1210 rows x 1 columns]       Down_Time
0      0.990272
1      0.525292
2      0.404669
3      0.058366
4      0.280156
...         ...
1205   0.031128
1206   0.015564
1207   0.212062
1208   0.387160
1209   0.120623

[1210 rows x 1 columns]


In [17]:
# Split data into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(X,y, train_size=0.7,test_size=0.3, random_state=42)

In [18]:
# Create a Random Forest regressor
rf = RandomForestRegressor(n_estimators=100, random_state=42)

In [19]:
# Train the classifier on the training data
rf.fit(X_train, y_train)

C:\Users\15878\AppData\Local\Temp\ipykernel_9004\4162563767.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf.fit(X_train, y_train)


RandomForestRegressor(random_state=42)

In [20]:
# Predict on the testing data
y_pred = rf.predict(X_test)

# Calculate the mean squared error
mse = mean_squared_error(y_test, y_pred)

print("Mean Squared Error:", mse)

Mean Squared Error: 0.08940203913433689


### Model Development

In [16]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [9]:
# Split data into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(X,y, train_size=0.7,test_size=0.3, random_state=42)

######  Here X = Availability is the input data which is a pandas DataFrame containing the features to be used for prediction, and y = Downtime is the target variable which is a pandas Series containing the corresponding target values.     Here X= Availability Y = Downtime

###### The train_test_split function is being used to split the data into training and testing subsets in the ratio of 70:30, where the train_size parameter is set to 0.7 and the test_size parameter is set to 0.3. random_state parameter is set to 42 to ensure the same random splits are obtained every time the code is run.

In [10]:
# Create a Random Forest regressor
rf = RandomForestRegressor(n_estimators=100, random_state=42)


####  This line of code initializes a random forest regression model with 100 decision trees (n_estimators=100) and sets the random state to 42. The random state is a parameter that determines the seed used by the random number generator. Setting this parameter ensures that the results are reproducible.

In [11]:
# Train the classifier on the training data
rf.fit(X_train, y_train)

C:\Users\15878\AppData\Local\Temp\ipykernel_12536\4162563767.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf.fit(X_train, y_train)


RandomForestRegressor(random_state=42)

#### This code trains a random forest regression model on the training data. The fit() method is used to train the model by fitting the model to the training data

In [12]:
# Predict on the testing data
y_pred = rf.predict(X_test)

# Calculate the mean squared error
mse = mean_squared_error(y_test, y_pred)

print("Mean Squared Error:", mse)

Mean Squared Error: 0.03947544738692804




#### The mean squared error (MSE) is a measure of the average squared difference between the predicted values and the actual values. The higher the MSE value, the worse the performance of the model.

#### In this case, the MSE value is 0.039, which is relatively low, indicating good performance of the model.

### Decision Tree

In [13]:
# Import libraries
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

In [14]:
# Load data
data = pd.read_csv(r"C:\Users\15878\DA\SGD_Data_Pred.csv")

data


idle_Cycle_time         HT          MT  Stop_time  Run_Time  Down_Time  \
0               145.0   0.000000    0.000000          0         0        509   
1                 0.0   0.000000    0.000000          0         0        270   
2               145.0   0.000000    0.000000          0         0        208   
3                 0.0   0.000000    0.000000          0         0         30   
4               145.0  27.500000  119.277778         20       196        144   
...               ...        ...         ...        ...       ...        ...   
1205            145.0  42.000000   90.000000          1        14         16   
1206            150.0  49.269231  130.730769         19       314          8   
1207            145.0  39.088235  176.058823         35       324        109   
1208            145.0  47.500000  117.576923         26       241        199   
1209            145.0  63.071429  143.392857         54       263         62   

      Total_Count  rejection  Good_Count  Operating_Time     Quality  \
0               0          0           0             509    0.000000   
1               0          0           0             270    0.000000   
2               0          0           0             208    0.000000   
3               0          0           0              30    0.000000   
4              72          4          68             358   94.444444   
...           ...        ...         ...             ...         ...   
1205            4          0           4              31  100.000000   
1206          118          4         114             340   96.610169   
1207          122          8         114             466   93.442623   
1208           98          8          90             465   91.836735   
1209          102          0         102             378  100.000000   

      Availability  Performance        OEE  
0         0.000000     0.000000   0.000000  
1         0.000000     0.000000   0.000000  
2         0.000000     0.000000   0.000000  
3         0.000000     0.000000   0.000000  
4        60.007446    80.961613  45.909578  
...            ...          ...        ...  
1205     47.120709    68.155112  32.115172  
1206     97.775486    88.895585  84.021226  
1207     70.332229    82.490091  59.137092  
1208     57.212986    88.996054  46.809426  
1209     83.666166    78.101072  65.344172  

[1210 rows x 14 columns]

In [15]:
data.corr()

idle_Cycle_time        HT        MT  Stop_time  Run_Time  \
idle_Cycle_time         1.000000  0.361301  0.336246   0.329991  0.306251   
HT                      0.361301  1.000000  0.541456   0.507875  0.261067   
MT                      0.336246  0.541456  1.000000   0.421145  0.428051   
Stop_time               0.329991  0.507875  0.421145   1.000000  0.541044   
Run_Time                0.306251  0.261067  0.428051   0.541044  1.000000   
Down_Time               0.125443 -0.093326 -0.143360  -0.078341 -0.130353   
Total_Count             0.354860  0.311833  0.486803   0.693066  0.749819   
rejection               0.115793  0.078085  0.103753   0.181064  0.177313   
Good_Count              0.353614  0.312814  0.489889   0.694603  0.753073   
Operating_Time          0.361610  0.220906  0.334977   0.514519  0.855355   
Quality                 0.298597  0.602097  0.718077   0.515380  0.520631   
Availability            0.208226  0.405264  0.548631   0.523826  0.576643   
Performance             0.019405 -0.007585 -0.010643  -0.058880 -0.061546   
OEE                     0.026204  0.011894  0.026827  -0.048234 -0.047943   

                 Down_Time  Total_Count  rejection  Good_Count  \
idle_Cycle_time   0.125443     0.354860   0.115793    0.353614   
HT               -0.093326     0.311833   0.078085    0.312814   
MT               -0.143360     0.486803   0.103753    0.489889   
Stop_time        -0.078341     0.693066   0.181064    0.694603   
Run_Time         -0.130353     0.749819   0.177313    0.753073   
Down_Time         1.000000    -0.146167   0.073835   -0.156088   
Total_Count      -0.146167     1.000000   0.325494    0.996714   
rejection         0.073835     0.325494   1.000000    0.247832   
Good_Count       -0.156088     0.996714   0.247832    1.000000   
Operating_Time    0.397888     0.638516   0.208325    0.636376   
Quality          -0.213405     0.627590   0.064652    0.637489   
Availability     -0.414485     0.682269   0.118090    0.688935   
Performance       0.012779    -0.067681  -0.004595   -0.068952   
OEE              -0.046283    -0.052423  -0.022939   -0.051747   

                 Operating_Time   Quality  Availability  Performance       OEE  
idle_Cycle_time        0.361610  0.298597      0.208226     0.019405  0.026204  
HT                     0.220906  0.602097      0.405264    -0.007585  0.011894  
MT                     0.334977  0.718077      0.548631    -0.010643  0.026827  
Stop_time              0.514519  0.515380      0.523826    -0.058880 -0.048234  
Run_Time               0.855355  0.520631      0.576643    -0.061546 -0.047943  
Down_Time              0.397888 -0.213405     -0.414485     0.012779 -0.046283  
Total_Count            0.638516  0.627590      0.682269    -0.067681 -0.052423  
rejection              0.208325  0.064652      0.118090    -0.004595 -0.022939  
Good_Count             0.636376  0.637489      0.688935    -0.068952 -0.051747  
Operating_Time         1.000000  0.387350      0.333418    -0.052241 -0.069913  
Quality                0.387350  1.000000      0.766496    -0.010061  0.043726  
Availability           0.333418  0.766496      1.000000    -0.025298  0.069640  
Performance           -0.052241 -0.010061     -0.025298     1.000000  0.443548  
OEE                   -0.069913  0.043726      0.069640     0.443548  1.000000

In [16]:
from sklearn.preprocessing import MinMaxScaler
data = pd.read_csv(r"C:\Users\15878\DA\SGD_Data_Pred.csv")
# create a MinMaxScaler object
scaler = MinMaxScaler()

# select columns to scale
cols_to_scale = ['Availability']
cols_to_scale1 = ['Down_Time']

# apply Min-Max scaling to selected columns
data[cols_to_scale] = scaler.fit_transform(data[cols_to_scale])
data[cols_to_scale1] = scaler.fit_transform(data[cols_to_scale1])
print(data[cols_to_scale],data[cols_to_scale1])

      Availability
0         0.000000
1         0.000000
2         0.000000
3         0.000000
4         0.600074
...            ...
1205      0.471207
1206      0.977755
1207      0.703322
1208      0.572130
1209      0.836662

[1210 rows x 1 columns]       Down_Time
0      0.990272
1      0.525292
2      0.404669
3      0.058366
4      0.280156
...         ...
1205   0.031128
1206   0.015564
1207   0.212062
1208   0.387160
1209   0.120623

[1210 rows x 1 columns]


In [17]:
# Split data into features and target variable
X = data[cols_to_scale]
y = data[cols_to_scale1]
print(X,y)

      Availability
0         0.000000
1         0.000000
2         0.000000
3         0.000000
4         0.600074
...            ...
1205      0.471207
1206      0.977755
1207      0.703322
1208      0.572130
1209      0.836662

[1210 rows x 1 columns]       Down_Time
0      0.990272
1      0.525292
2      0.404669
3      0.058366
4      0.280156
...         ...
1205   0.031128
1206   0.015564
1207   0.212062
1208   0.387160
1209   0.120623

[1210 rows x 1 columns]


In [18]:
X_train, X_test, y_train, y_test = train_test_split(X,y,train_size=0.7,test_size=0.3, random_state=42)

In [19]:
# Create a Decision Tree Regressor
dtr = DecisionTreeRegressor(max_depth=5, random_state=42)

In [20]:
# Train the regressor on the training data
dtr.fit(X_train, y_train)

DecisionTreeRegressor(max_depth=5, random_state=42)

In [21]:
y_pred = dtr.predict(X_test)

In [22]:
mse = mean_squared_error(y_test, y_pred)
print("Mean Squared Error:", mse)

Mean Squared Error: 0.03453775239880787


## Inference:

#### - The model was able to predict the target variable (Down_Time) based on the feature variable (Availability).
#### - The mean squared error of the prediction was 0.034, which indicates that the model has a good fit to the data.
#### - Since the model is a Decision Tree Regressor, we can interpret the results by analyzing the tree structure. The tree can help identify the most important features that are driving the predictions and provide insights into how the model is making decisions.

### In this case, the MSE of 0.0395 is slightly higher than the MSE of the Decision Tree Regressor model with max_depth=5, which had an MSE of 0.0345. This suggests that the Decision Tree Regressor model might be slightly more accurate in predicting the target variable than the Random Forest Regressor model